ovh/resource_ovh_iploadbalancing_tcp_farm_server.go

```diff
@@ -0,0 +1,234 @@
+	Weight               *int    `json:"weight"`
+	Probe                *bool   `json:"probe"`
+	Ssl                  *bool   `json:"ssl"`
+	Backup               *bool   `json:"backup"`
+	Status               *string `json:"status"`
+}
+
+func resourceIpLoadbalancingTcpFarmServer() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceIpLoadbalancingTcpFarmServerCreate,
+		Read:   resourceIpLoadbalancingTcpFarmServerRead,
+		Update: resourceIpLoadbalancingTcpFarmServerUpdate,
+		Delete: resourceIpLoadbalancingTcpFarmServerDelete,
+		Schema: map[string]*schema.Schema{
+			"service_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"farm_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"display_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"address": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					ip := v.(string)
+					if net.ParseIP(ip).To4() == nil {
+						errors = append(errors, fmt.Errorf("Address %s is not an IPv4", ip))
+					}
+					return
+				},
+			},
+			"ssl": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"cookie": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"port": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"proxy_protocol_version": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					err := validateStringEnum(v.(string), []string{"v1", "v2", "v2-ssl", "v2-ssl-cn"})
+					if err != nil {
+						errors = append(errors, err)
+					}
+					return
+				},
+			},
+			"chain": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"weight": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+			},
+			"probe": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"backup": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"status": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					err := validateStringEnum(v.(string), []string{"active", "inactive"})
+					if err != nil {
+						errors = append(errors, err)
+					}
+					return
+				},
+			},
+		},
+	}
+}
+
+func resourceIpLoadbalancingTcpFarmServerCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	newBackendServer := &IpLoadbalancingTcpFarmServer{
+		DisplayName:          getNilStringPointer(d.Get("display_name").(string)),
```

Comment:
i don't see the use of "getNilStringPointer"

AFAIK, there's no such thing elswhere in terraform providers. what are we trying to do here ?

Reply:
assuming "display_name" is not set, `d.Get("display_name").(string)` returns an empty string `""`, but a string none the less. gos `json.Marshal` method correctly interprets it as an empty string rather then a *not set* string. This is why the struct uses pointers instead of string values, so they can have `nil` value. getNilStringPointer then gets the string value and turns it into a pointer to string value, that in case of `len(<string>) == 0` returns nil, so that `json.Marshal()` instead of generating `"displayName": ""` generates `"displayName": null` which OVH API expects (ie. it will complain if you will have a zero value field instead of null for attributes that can not be changed after resource creation

Reply:
have you tried with d.Get("display_name").(*string) ?

Reply:
in fact I did :) it did not work. Don't think type assertion is meant to work like that at all, and even if it did, it would have no way to distinct between nil and zero value, so instead of recovering `nil`, it would recover `*""` but as I said, it's not doing even that.

Reply:
the functions I provided here do just a bit more then what you can see in Pointer retrievals for aws like : 
https://github.com/terraform-providers/terraform-provider-aws/blob/80e4ae53941c6fc87239f1d704e2dfe6a4dec3b6/aws/resource_aws_mq_broker.go#L185
https://github.com/aws/aws-sdk-go/blob/71a68668c287f9ad58d0c58c78e27f08eb39fa49/aws/convert_types.go#L5